core/controllers/classifier.py

```diff
@@ -144,6 +142,75 @@ def post(self):
 
         return self.render_json({})
 
+    @acl_decorators.open_access
+    def get(self):
+        """Handles GET requests.
+
+        Retrieves the name of the file on GCS storing the trained model
+        parameters and transfers it to the frontend.
+        """
+        exploration_id = self.request.get('exploration_id')
+        state_name = self.request.get('state_name')
+
+        try:
+            exp_version = int(self.request.get('exploration_version'))
+            exploration = exp_fetchers.get_exploration_by_id(
+                exploration_id, version=exp_version)
+            interaction_id = exploration.states[state_name].interaction.id
+        except:
+            raise self.InvalidInputException(
+                'Entity for exploration with id %s and version %s not found.'
+                % (exploration_id, self.request.get('exploration_version')))
+
+        if interaction_id not in feconf.INTERACTION_CLASSIFIER_MAPPING:
+            raise self.PageNotFoundException(
+                'No classifier algorithm found for %s interaction' % (
+                    interaction_id))
+
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        algorithm_version = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_version']
+
+        state_training_jobs_mapping = (
+            classifier_services.get_state_training_jobs_mapping(
+                exploration_id, exp_version, state_name))
+        if state_training_jobs_mapping is None:
+            raise self.InvalidInputException
```

Comment:
These are not user facing exceptions. There is another service for this which gracefully handles all errors. None of this actually affects any part of UI.

Reply:
This `InvalidInputException` is in the controller file so I'm expecting it to be a user-facing exception. (In general, InvalidInputException is a user-facing exception)
![image](https://user-images.githubusercontent.com/16653571/90323423-e0496080-df7e-11ea-8a4f-e1489c90ac55.png)


You can fix this in your next PR! (Make sure to create an issue for the same!)

Reply:
@DubeySandeep, I think there is misunderstanding on my side here. What do you mean by phrase "user-facing exception". This controller is not intended to be invoked by end user directly, there is no action on  Oppia UI that can lead to its invocation. A separate service will use this controller and handle any exceptions thrown by the controller. At this point, only the type of exception is useful for this service. Hence, I am not sure what you mean by "user-facing exception" here. What should have been different from what I have written? Also, I have added test cases for this but they are not of type `self.assertRaisesRegexp`, instead they make an http request and their the error status code is matched (through callback functions).

Reply:
Hi @prasanna08 -- I think one concern here (which is actually an issue I had when reviewing your PR too) is that this looks like a user-facing controller and there currently isn't a way to distinguish it from a non-user-facing controller (in fact you have labelled it "open access"). Even if there is no way to access it by the GUI, users can still make a wget/curl call to the URL in question.

We may need to introduce the concept of VM handlers that can only be accessed by services in Oppia VMs, otherwise the default assumption is that anything in core/controllers is accessible by an end user. Do you have any thoughts on how best to do this? Would using a different decorator be sufficient? If not then perhaps we need to cordon such handlers off into a separate module.

Reply:
I see. Actually, there are two types of handles in this file:
1. Accessed by VM: These are `TrainedClassifierHandler.post` and `NextJobHandler.get`. We have necessary security and authentication mechanism such as `VM_ID` and `SECRET_KEY` which make sure that only VM can get data, everyone else will get an error.
2. Accessed by Oppia frontend: There is only one `TrainedClassifierHandler.get`. Now this is indeed open access since it is not accessed by VM but Oppia's frontend stack to get the trained classifier handler. 

Now, we can have a separate decorator for [1], where we move the signature verification code to a separate decorator, as it is common for both of these handlers. But not sure what to do about [2]. We can restrict access to [2] similar to whatever we do for get exploration (I think we have `acl_decorators.can_play_exploration`? Not sure) but yes, end users can still make requests to [2] and get filename of GCS datastore (but not the actual file, as GCS access has its own acl support). So, I can do changes for [1] but what do you suggest for [2]?

Reply:
Hm, yep, I think you should be able to just leave (2) to be similar to "get exploration", just like all the other handlers that are accessible to the frontend. Does that sound good?

Reply:
Ok, will do. Thanks.

Reply:
Hi @seanlip, even if the verification code is similar, there is going to be a design issue for implementing it as a part of separate decorator. The code would look something like below:
```
@acl_decorators.oppia_ml_access(handler_type='TrainedClassifierHandler')
def post(self):
```
The reason for this is that one handler receives a proto binary from Oppia-ML while other receives a simple json and hence we need to make some distinction to be able to identify the appropriate method to decode the data. The `handler_type` is just one example of how it can be done.

However, this design is not scalable at all as for every handler which uses different structure from the rest. Hence, I am still leaning towards current design which is better from maintainability perspective. What do you think?

Reply:
Sorry, I'm not sure I understand. What is your proposal for how to proceed?

If it's leaving things as the status quo, I don't think that works, for the reasons described earlier in this thread...

Reply:
There is another way of doing it: have another function which decodes the incoming object and returns necessary components like `message` and `vm_id`. This results in usage of `self` from the decorator function which is common case for a few decorators as I saw in `acl_decorators.py`. This can alleviate requirements of having an extra handler_type from decorator arguments.

Reply:
Hm, OK. I don't fully follow the suggestion yet, but nothing bad stands out to me atm. Perhaps you could implement it and let's see how it looks?